File: apps/api/src/jobs/fraudDetection.ts
import cron from 'node-cron';
import { query } from '../config/database';
import { notifyFraudAlert } from '../services/sms.service';

/**
 * Fraud detection job - runs every 30 minutes
 * Checks for suspicious transaction patterns
 */
export function startFraudDetectionJob(): void {
  cron.schedule('*/30 * * * *', async () => {
    try {
      // Pattern 1: Multiple failed transactions in short period
      const failedPattern = await query(
        `SELECT user_id, COUNT(*) as failed_count, MAX(phone) as phone
         FROM transactions
         WHERE status = 'failed' AND created_at >= NOW() - interval '1 hour'
         GROUP BY user_id
         HAVING COUNT(*) >= 5`
      );

      for (const alert of failedPattern.rows) {
        const existing = await query(
          `SELECT id FROM fraud_alerts 
           WHERE user_id = $1 AND alert_type = 'multiple_failures' AND created_at >= NOW() - interval '2 hours'`,
          [alert.user_id]
        );

        if (existing.rows.length === 0) {
          await query(
            `INSERT INTO fraud_alerts (user_id, alert_type, severity, description)
             VALUES ($1, 'multiple_failures', 'high', $2)`,
            [alert.user_id, `${alert.failed_count} failed transactions in the last hour`]
          );

          // Get user phone
          const userResult = await query('SELECT phone FROM users WHERE id = $1', [alert.user_id]);
          if (userResult.rows[0]) {
            notifyFraudAlert(userResult.rows[0].phone, 'Multiple failed transactions detected on your account').catch(console.error);
          }
        }
      }

      // Pattern 2: Unusually large transaction
      const largePattern = await query(
        `SELECT t.user_id, t.id as transaction_id, t.amount, u.phone
         FROM transactions t
         JOIN users u ON t.user_id = u.id
         WHERE t.created_at >= NOW() - interval '30 minutes'
           AND t.amount > (
             SELECT COALESCE(AVG(amount) * 5, 100000)
             FROM transactions
             WHERE user_id = t.user_id AND status = 'completed'
           )`
      );

      for (const alert of largePattern.rows) {
        const existing = await query(
          `SELECT id FROM fraud_alerts WHERE transaction_id = $1`,
          [alert.transaction_id]
        );

        if (existing.rows.length === 0) {
          await query(
            `INSERT INTO fraud_alerts (user_id, transaction_id, alert_type, severity, description)
             VALUES ($1, $2, 'large_transaction', 'medium', $3)`,
            [alert.user_id, alert.transaction_id, `Unusually large transaction of KES ${parseFloat(alert.amount).toLocaleString()}`]
          );
        }
      }

      // Pattern 3: Rapid successive transactions
      const rapidPattern = await query(
        `SELECT user_id, COUNT(*) as tx_count, MAX(phone) as phone
         FROM transactions
         WHERE created_at >= NOW() - interval '5 minutes'
         GROUP BY user_id
         HAVING COUNT(*) >= 10`
      );

      for (const alert of rapidPattern.rows) {
        const existing = await query(
          `SELECT id FROM fraud_alerts
           WHERE user_id = $1 AND alert_type = 'rapid_transactions' AND created_at >= NOW() - interval '30 minutes'`,
          [alert.user_id]
        );

        if (existing.rows.length === 0) {
          await query(
            `INSERT INTO fraud_alerts (user_id, alert_type, severity, description)
             VALUES ($1, 'rapid_transactions', 'critical', $2)`,
            [alert.user_id, `${alert.tx_count} transactions in 5 minutes — possible automated attack`]
          );

          const userResult = await query('SELECT phone FROM users WHERE id = $1', [alert.user_id]);
          if (userResult.rows[0]) {
            notifyFraudAlert(userResult.rows[0].phone, 'Unusual rapid transactions detected').catch(console.error);
          }
        }
      }
    } catch (error) {
      console.error('Fraud detection error:', error);
    }
  });

  console.log('🛡️ Fraud detection job scheduled (every 30 minutes)');
}

File: OneDrive/Desktop/SharedProj/HazinaHub/backend/app/jobs/fraud_detection.py
from datetime import datetime, timedelta
from apscheduler.schedulers.asyncio import AsyncIOScheduler

from app.models.cashflow import Cashflow
from app.models.fraud_alert import FraudAlert
from app.models.user import User
from app.services.sms import notify_fraud_alert


async def run_fraud_detection():
    """Fraud detection — runs every 30 minutes."""
    try:
        now = datetime.utcnow()
        one_hour_ago = now - timedelta(hours=1)
        thirty_min_ago = now - timedelta(minutes=30)
        five_min_ago = now - timedelta(minutes=5)

        # ── Pattern 1: Multiple failed cashflows in 1 hour ───────────
        pipeline_failed = [
            {"$match": {"status": "failed", "created_at": {"$gte": one_hour_ago}}},
            {"$group": {"_id": "$user_id", "count": {"$sum": 1}, "phone": {"$last": "$phone"}}},
            {"$match": {"count": {"$gte": 5}}},
        ]
        failed_alerts = await Cashflow.aggregate(pipeline_failed).to_list()
        for alert in failed_alerts:
            two_hrs_ago = now - timedelta(hours=2)
            existing = await FraudAlert.find_one(
                FraudAlert.user_id == alert["_id"],
                FraudAlert.alert_type == "multiple_failures",
                FraudAlert.created_at >= two_hrs_ago,
            )
            if not existing:
                await FraudAlert(
                    user_id=alert["_id"],
                    alert_type="multiple_failures",
                    severity="high",
                    description=f"{alert['count']} failed cashflow attempts in the last hour",
                ).insert()
                if alert.get("phone"):
                    try:
                        await notify_fraud_alert(
                            alert["phone"],
                            "Multiple failed cashflow attempts detected on your account",
                        )
                    except Exception:
                        pass

        # ── Pattern 2: Unusually large cashflow (5× avg) ─────────────
        recent_txs = await Cashflow.find(
            Cashflow.created_at >= thirty_min_ago,
        ).to_list()

        for tx in recent_txs:
            # Calculate this user's avg
            pipeline_avg = [
                {"$match": {"user_id": tx.user_id, "status": "completed"}},
                {"$group": {"_id": None, "avg": {"$avg": "$amount"}}},
            ]
            avg_result = await Cashflow.aggregate(pipeline_avg).to_list()
            user_avg = avg_result[0]["avg"] if avg_result else 100000

            if tx.amount > user_avg * 5:
                existing = await FraudAlert.find_one(
                    FraudAlert.cashflow_id == tx.id,
                )
                if not existing:
                    await FraudAlert(
                        user_id=tx.user_id,
                        cashflow_id=tx.id,
                        alert_type="large_transaction",
                        severity="medium",
                        description=f"Unusually large cashflow of KES {tx.amount:,.0f}",
                    ).insert()

        # ── Pattern 3: Rapid successive cashflows (≥10 in 5 min) ─────
        pipeline_rapid = [
            {"$match": {"created_at": {"$gte": five_min_ago}}},
            {"$group": {"_id": "$user_id", "count": {"$sum": 1}, "phone": {"$last": "$phone"}}},
            {"$match": {"count": {"$gte": 10}}},
        ]
        rapid_alerts = await Cashflow.aggregate(pipeline_rapid).to_list()
        for alert in rapid_alerts:
            existing = await FraudAlert.find_one(
                FraudAlert.user_id == alert["_id"],
                FraudAlert.alert_type == "rapid_transactions",
                FraudAlert.created_at >= thirty_min_ago,
            )
            if not existing:
                await FraudAlert(
                    user_id=alert["_id"],
                    alert_type="rapid_transactions",
                    severity="critical",
                    description=f"{alert['count']} cashflows logged in 5 minutes — possible automated attack",
                ).insert()
                if alert.get("phone"):
                    try:
                        await notify_fraud_alert(
                            alert["phone"],
                            "Unusual rapid cashflows detected",
                        )
                    except Exception:
                        pass

    except Exception as e:
        print(f"Fraud detection error: {e}")


def register_fraud_detection(scheduler: AsyncIOScheduler):
    scheduler.add_job(run_fraud_detection, "interval", minutes=30, id="fraud_detection")
    print("[SCHEDULED] Fraud detection job scheduled (every 30 minutes)")
